### DISCUSSION TITLE
Why hybrid ASR could train well with conformer but not fastconformer (everything else being the same)?

### DISCUSSION THREAD

**Discussion by bene-ges** [NVIDIA-NeMo/NeMo#7797] [Q&A]
Hi,
I tried to train from scratch two hybrid models on microdataset of 900 sentences using char tokens. 
The hyperparameters were taken from default nemo config `conformer/conformer_transducer_char.yaml`
The single difference was using conformer/fastconformer in them. The first model converges well, the second doesn't. Can anyone give a hint - why could it be?

![image](https://github.com/NVIDIA/NeMo/assets/61418381/d63de232-ce9d-4fe9-bc0c-d412e68a5932)

**Comment by titu1994** [comment]
Are both models using hybrid loss but just encoder changed? Depending on seq length, CTC might not work well. Fast Conformer is 8x stride model, please make sure your transcript is always at least slightly longer than the audio seq/ 8. 

Rnnt loss conveged well but CTC loss diverged, you can try using smaller CTC aux weight (0.1 instead of 0.3 which is the default). 

The hyper parameters are tuned for pretty large datasets, so they might not fit am overfitting task of 900 samples.

**Comment by bene-ges** [comment]
thanks, @titu1994,
 yes, same hybrid loss, only encoder params(subsampling and conv_kernel_size) correspond to fastconformer.

I tried CTC aux weight=0.1, 
it became slightly better, ctc val loss doesn't diverge anymore
but still very far from simple(not fast) conformer. Is it that training ctc with fastconformer requires more iterations? 

![image](https://github.com/NVIDIA/NeMo/assets/61418381/c82495da-47e8-4587-9cd4-ebabc13ba380)

**Comment by bene-ges** [comment]
I also tried to modify encoder to support fastconformer+streaming+multi like in your model `stt_en_fastconformer_hybrid_large_streaming_multi`.
Result: CTC doesn't train at all - even train loss doesn't decrease. Look at pink graph. Is it expected behaviour?
![image](https://github.com/NVIDIA/NeMo/assets/61418381/fac87e4c-c92c-4904-bdba-a226c758786d)

**Reply by bmwshop** [reply]
Have you tried just training the Fast Conformer CTC model separately? Maybe that will offer a clue. I don't believe we attempted to train it with char tokens.  As @titu1994 mentioned,  CTC models won't work if the transcripts are very short, the total # of output tokens less than the 8x reduced (in case of FC, in case of C, 4x) audio sequence.  It would be good to rule this out.

**Comment by bene-ges** [comment]
@bmwshop, I do not quite understand the condition about 8x wrt audiosequence. I printed signal_len and transcript_len that come from dataloader, average values are like 118000 for signal_len, 65 for transcript_len.

Tried to train ctc models separately.
Similar picture: fastconformer converges much slower, fastconformer+streaming even worse.

![image](https://github.com/NVIDIA/NeMo/assets/61418381/18d572fb-378e-4a2e-9866-a3317c485ce6)

**Comment by bene-ges** [comment]
Returning to hybrid model, now using bpe instead of char. Same micro dataset as in previous experiments.
hybrid + fastconformer + streaming + multi, ctc_weight=0.1
If I try to overfit from scratch - ctc doesn't seem to train.
But if I initialize from your pretrained model `stt_en_fastconformer_hybrid_large_streaming_multi`, it works ok, both losses go down. I use different language, so replaced tokenizer in both cases.

![image](https://github.com/NVIDIA/NeMo/assets/61418381/228560e4-cd1a-48a1-9513-ce945d9be052)
